### PR TITLE
nixos-rebuild: init as package

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Move the nixos-rebuild script from the nixos distribution into the
package set.

This allows non-nixos users to run the script as well. For example,
deploying a remote machine with:

    nixos-rebuild --target-host root@hostname --build-host root@hostname

If that idea is interesting, potentially move all the other nixos
commands to packages.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by zimbatm** [comment]
Seems like it's popular so I moved all the other tools.

There is a new `<nixpkgs/.codeName>` file so it's accessible from the pkgs side. I think we should introduce a `<nixpkgs/version.nix>` that holds the structured data but wanted to avoid breaking backward-compatibility.

`nixos-rebuild-vms` depends on `<nixpkgs/nixos/modules/installer/tools/nixos-build-vms/build-vms.nix>` which is not ideal in terms of data locality but I wanted to avoid breaking backward-compatibility here.

**Comment by zimbatm** [comment]
@GrahamcOfBorg build nixos-build-vms nixos-enter nixos-generate-config nixos-install nixos-option nixos-rebuild nixos-version

**Comment by Mic92** [comment]
It seems that one could replace nixops with a shell script for the most part.

**Comment by zimbatm** [comment]
I want to back-port this to 18.09. Did a bunch of tests and it seems to be working well.

Any objections?

**Comment by zimbatm** [comment]
@GrahamcOfBorg build nixos-build-vms nixos-enter nixos-generate-config nixos-install nixos-option nixos-rebuild nixos-version

**Comment by zimbatm** [comment]
I'll just go ahead and merge this in master. If everything goes well, I will cherry-pick this in 18.09 next Monday.

File: .codeName [hunk 1]
```diff
@@ -0,0 +1 @@
+Koi
```
Comment (edolstra) [inline .codeName]: Please don't use camelCase filenames. Also, does this really have to be in the root directory of the repository?

**Comment by edolstra** [comment]
I'm strongly against this. These are not packages - they should be in the NixOS source tree, not in `pkgs`.

File: .codeName (hunk 1 above)
Comment (zimbatm) [reply .codeName]: how about something like this https://github.com/NixOS/nixpkgs/pull/48588 ?

**Comment by zimbatm** [comment]
@edolstra is it a question of data locality, so that the script can evolve at the same pace as NixOS?

Personally I don't tend to make a bit distinction between both as they are in the same repo and have the same release schedule. NixOS is just another type of (albeit complex) derivation output.

**Comment by Mic92** [comment]
I would even say it is a complex derivation.

**Comment by Ekleog** [comment]
@zimbatm At first I agreed with the idea that it'd be nice to have `nixos-rebuild` be a regular package, but it's true it makes sense only in the NixOS context, and thus has no meaning on all other systems.

So it'd require, to make sense, to either add a `platforms.nixos` platform, which sounds overkill… or to live in the `nixos/` tree.

Now, maybe a NixOS overlay to nixpkgs could be done, containing these? That'd solve both issues, with both the scripts as actual packages and packages in `nixos/`, explicitly being marked for NixOS.

**Comment by zimbatm** [comment]
I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem.

After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.

**Comment by Ekleog** [comment]
> I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem.
>
> After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.

Hmm… why not a nixos overlay put in the `nixos/` folder and enabled by
default on NixOS but pick-able on-demand on other systems? That would
sound like it'd also solve all the problems caused by `nixos-*` tools
not being packaged, to me

**Comment by Mic92** [comment]
> I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem. After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.
> Hmm… why not a nixos overlay put in the `nixos/` folder and enabled by default on NixOS but pick-able on-demand on other systems? That would sound like it'd also solve all the problems caused by `nixos-*` tools not being packaged, to me

I think that can be achieved by using `(nixos{})` already.

**Comment by Mic92** [comment]
* easily boostrap nixos from a linux distribution + nix.

**Comment by marler8997** [comment]
I'm currently installing nix onto my second hard drive from Ubuntu.  I got the drive partitioned and the filesystems mounted, but then I was surprised to find that `nixos-generate-config` was not available as a package...?  Since this was removed...what's the recommended way forward for me here to complete the install?  Note that I don't have a CD drive.

**Comment by Mic92** [comment]
If you have USB port/stick, use dd the image to that. You can also use an kexec image that can drop you into a installer from your existing Linux: https://github.com/nix-community/nixos-generators

**Comment by marler8997** [comment]
Thanks for the tips, I didn't know about kexec I'll have to learn about that one.

So `nixos-generate-config` is still in nixpkgs, it's just not a top-level entry anymore.  Seems odd to a new Nix user like me.  I would think the reason not to put a tool in the top-level attribute set would be if it had a name that could clash...but `nixos-generate-config` is pretty explicit.

**Pull request merged**